core/src/impl/Kokkos_ViewWithProperty.hpp

```diff
@@ -0,0 +1,147 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 2.0
+//              Copyright (2014) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_VIEWWITHPROPERTY_HPP
+#define KOKKOS_VIEWWITHPROPERTY_HPP
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename DataType, typename ArrayLayout, typename DeviceType, typename MemoryTraits>
+struct OrderedViewProperties
+{
+  using data_type = DataType;
+  using device_type = DeviceType;
+  using array_layout = ArrayLayout;
+  using memory_traits = MemoryTraits;
+};
+
+template <typename OrderedProps>
+using view_with_ordered_props_t = Kokkos::View<
+  typename OrderedProps::DataType,
+  typename OrderedProps::ArrayLayout,
+  typename OrderedProps::DeviceType,
+  typename OrderedProps::MemoryTraits>;
+
+template <typename ViewType>
+using ordered_props_of_view_t = OrderedViewProperties<
```

Comment:
It kinda feels like all this "canonicalize View properties" machinery should go in one place in one file, so that it's easier to maintain if properties should change.  Also, do we still do the specialization stuff for dynamic FAD types?

Reply:
There is a file for @crtrott 's flavor of canonicalization, but he goes a bit too far in that his system converts compile-time dimensions into runtime dimensions. I don't think this function should do that. FAD types are sadly still magical, but I don't think this code needs to know about that magic at all. This is all above the level of Kokkos::View.